### ISSUE TITLE
Hardware acceleration causing troubles in xbox

### ISSUE THREAD

**Issue by villagra** [ffmpeginteropx/FFmpegInteropX#242]
Hi, I've started to get numerous reports about green video the video freezes on xbox. I got a hint that the new D3D11 acceleration might be the cause....  I released an update not so long ago updating to the latest nuget packages and seems like something in the ffmpeg build is causing that. 
Is it possible to disable it via configuration?
Thank you!

**Comment by brabebhin** [comment]
There are some config options to disable it. VideoDecoderMode i think it was.

**Comment by villagra** [comment]
Thank you 👍

**Comment by brabebhin** [comment]
Did you manage to reproduce the issue?

**Comment by villagra** [comment]
Yes! I did reproduce it using a xbox one X (it doesn't happen with the new series X). 
Setting the Config.VideoDecoderMode = VideoDecoderMode.AutomaticSystemDecoder has fixed the problem.

Thanks

**Comment by brabebhin** [comment]
Well if hardware support wasn't available, then it should have switched to software decoding. I don't have an xbox one X, would it be possible for you to debug and see why it doesn't auto switch to software?

**Comment by villagra** [comment]
Unfortunately my c++ experience is very very limited...  😒

**Comment by TheElixZammuto** [comment]
Hi, I have the same problem on another application that uses FFMpeg (but not this distribution) on UWP. I have video freezes on Xbox but not on Windows. So maybe a problem relative to FFmpeg itself?

**Comment by brabebhin** [comment]
Hardware acceleration is device dependent. Do you have a video file which reproduces the issues?

**Comment by TheElixZammuto** [comment]
Unfortunately I don't have a video file since I'm streaming from a custom protocol that gives me H264 NAL. Also unfortunately since I'm not part of any program, UWP is the only solution for me for development on Xbox. I'm going to downgrade FFMpeg and let you know if this does resolve the issue

PS: I'm not using FFMpegInteropX, but I found this thread and would be useful to check out together what's causing this

**Comment by brabebhin** [comment]
This copying thing we had in the beginning, but we noticed it crashed the performance of 4K H265 videos, so we gave up on that and instead used the more streamlined device to device memory copy. 
Unless I can get a way to reproduce this, I will unfortunately be unable to help you with this. But I think it might be worth reporting the problem to Microsoft. This really does sound like a problem with Xbox environment, rather than the framework.

If you can provide a PR with feature flag which toggles the system memory copy then we could integrate this into repository (Idealy by sub-classing the existing hardware sample provider). Unfortunately I will not be able to participate in large coding for this project for the next 2-3 months, maybe @lukasf can help. But I could take a look if someone else does most of the heavy lifting.

**Comment by lukasf** [comment]
@TheElixZammuto I'm glad to hear you found a workaround that fixes your issues. In the first version using FFmpeg HW acceleration, we also used a copy approach (although I think it was direct GPU to GPU copy?). For some codecs, FFmpeg needs the hardware device before playback starts (and as you know, in UWP we can only get the device instance on playback start). But then we found ways to work around this issue on the PC, achieving much better performance on 4K material without any copy operations. It might be that our approach does not work on the older xbox due to some reason.

Without access to the actual xbox hardware, it is impossible for me to debug into this, to find out the real root cause. And I won't try to solve this blindly, without knowing what is really happening here and how the behavior is on xbox. If someone finds out how to properly fix this in FFmpegInteropX, feel free to create a PR.

Also, we have a good workaround of using system decoders, which should give excellent performance as well.
